### PR TITLE
Use CProverString.format as interface to the solver builtin function

### PR DESCRIPTION
**Pull request by romainbrenguier**
The way `String.format` is implemented in the string preprocessing is hacky and buggy. 
By adding a `CProverString.format` function as interface, the preprocessing becomes very simple, the complexity is then transfered to the java models, for which it is much easier to write clean code (than goto-program for java).
This should go together with an update of the models:  https://github.com/diffblue/java-models-library/pull/20

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by smowton — APPROVED** [review]
Seems sensible

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e74d75397292e8cb10ae183daede0c0bf103b5f0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106699730
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 660b193faa6fcc19d7e0491afd47d435638326a2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106716601
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: fe27d9984e6b5c0ba19d908e3a0e0ff317718957).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106736692
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: c577ab327ff8ed7b9c5a47132f20e790180c120c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106737115
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

File: src/solvers/strings/string_constraint_generator_format.cpp [hunk 1]
```diff
@@ -267,61 +268,42 @@ get_component_in_struct(const struct_exprt &expr, irep_idt component_name)
...
-  const typet &index_type,
-  const typet &char_type,
-  array_poolt &array_pool,
-  const messaget &message,
-  const namespacet &ns)
-{
+    symbol_generatort &fresh_symbol, const format_specifiert &fs,
+    const std::function<exprt(const irep_idt &)> &get_arg,
+    const typet &index_type, const typet &char_type, array_poolt &array_pool,
+    const messaget &message, const namespacet &ns) {
```
Comment (allredj) [inline src/solvers/strings/string_constraint_generator_format.cpp]: Why all the weird reformatting?

File: src/solvers/strings/string_constraint_generator_format.cpp [hunk 2]
```diff
@@ -299,7 +299,8 @@ add_axioms_for_format_specifier(
     return {res, std::move(return_code.second)};
   case format_specifiert::STRING:
   {
-    auto string_expr = get_string_expr(array_pool, get_arg("string_expr"));
+    const exprt arg_string = get_arg("string_expr");
```
Comment (allredj) [inline src/solvers/strings/string_constraint_generator_format.cpp:330]: :pick: typo in commit message ("functtion"). Also there's a missing "of" I think.

File: jbmc/regression/jbmc-strings/StringFormat/test-hex1.desc [hunk 3]
```diff
@@ -0,0 +1,6 @@
+CORE
+Test.class
+--function Test.testHex --cp `../../../../scripts/format_classpath.sh . ../../../lib/java-models-library/target/core-models.jar` --property "java::Test.testHex:(I)Ljava/lang/String;.assertion.1"
+^EXIT=10$
+^SIGNAL=0$
+line 7 assertion at file Test.java .*: FAILURE
```
Comment (allredj) [inline jbmc/regression/jbmc-strings/StringFormat/test-hex1.desc]: Line numbers don't seem to match. Assertions are at lines 5, 7, 9.

File: src/solvers/strings/string_constraint_generator_format.cpp [hunk 4]
```diff
@@ -363,21 +365,30 @@ exprt format_arg_from_string(const array_string_exprt &string,
         bitor_exprt{shl_exprt{typecast_exprt{string[2], type}, 16},
                     typecast_exprt{string[3], type}}};
   }
+
+  const and_exprt is_null{
+      equal_exprt{string[0], from_integer('n', string[0].type())},
+      equal_exprt{string[1], from_integer('u', string[0].type())},
+      equal_exprt{string[2], from_integer('l', string[0].type())},
+      equal_exprt{string[3], from_integer('l', string[0].type())},
+  };
```
Comment (allredj) [inline src/solvers/strings/string_constraint_generator_format.cpp]: :no_entry_sign: I think there's a missing constraint on the length of the string.

**Review by allredj — CHANGES_REQUESTED** [review]

File: src/solvers/strings/string_constraint_generator_format.cpp (hunk 1 above)
Comment (romainbrenguier) [reply src/solvers/strings/string_constraint_generator_format.cpp]: my clang-format is broken :sob:

Comment (tautschnig) [reply src/solvers/strings/string_constraint_generator_format.cpp]: Did #4461 possibly trip it up as you may need to move to a newer version?

Comment (romainbrenguier) [reply src/solvers/strings/string_constraint_generator_format.cpp]: Oh yes that looks related: 
```
.git-rewrite/t/.clang-format: Invalid argument
YAML:63:21: error: unknown key 'IndentPPDirectives'
IndentPPDirectives: AfterHash
```
I'm using clang-format-6.0

**Comment by romainbrenguier** [comment]
@allredj ready for rereview

File: src/solvers/strings/string_constraint_generator_format.cpp (hunk 1 above)
Comment (tautschnig) [reply src/solvers/strings/string_constraint_generator_format.cpp]: That seems surprising, it's been added in `clang-format` 6 (and `clang-format-6.0` works for me). But you might also try upgrading to `clang-format` 7, which is what we use on Travis.

File: jbmc/src/java_bytecode/java_string_library_preprocess.cpp [hunk 5]
```diff
@@ -1123,57 +1123,37 @@ code_blockt java_string_library_preprocesst::make_assign_function_from_call(
...
+    const irep_idt &function_id, symbol_table_baset &symbol_table,
+    code_blockt &code) {
   optionalt<struct_tag_typet> object_type;
 
   typet value_type;
-  if(type_name==ID_boolean)
-  {
-    value_type=java_boolean_type();
+  if (type_name == ID_boolean) {
+    value_type = java_boolean_type();
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/java_string_library_preprocess.cpp]: Still a lot of reformatting in here?!

Comment (romainbrenguier) [reply jbmc/src/java_bytecode/java_string_library_preprocess.cpp]: should be better now

File: src/solvers/strings/string_constraint_generator_format.cpp (hunk 1 above)
Comment (romainbrenguier) [reply src/solvers/strings/string_constraint_generator_format.cpp]: it's not available on ubuntu 16.04

**Review by allredj — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 211fd5ed5a45a7340ccd0dbcf1410dd5d7e59939).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106879072

File: src/solvers/strings/string_constraint_generator_format.cpp (hunk 1 above)
Comment (tautschnig) [reply src/solvers/strings/string_constraint_generator_format.cpp]: http://apt.llvm.org/ has repositories for Xenial.

File: jbmc/src/java_bytecode/java_string_library_preprocess.cpp (hunk 5 above)
Comment (tautschnig) [reply jbmc/src/java_bytecode/java_string_library_preprocess.cpp]: Hmm, then the changes are actually being reverted in a later commit - this still looks very much broken.

Comment (tautschnig) [reply jbmc/src/java_bytecode/java_string_library_preprocess.cpp]: It's "Remove code that is now unreachable" that removes the mis-formatted code.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 06e7e9f03892c128c7c62e9ecc12f76cd3065070).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106916333
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 7a920a142d350a4387228ae51f0c1ed220fb6c37).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106919141
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e45a2d0aba7f31b3e383c02679c72aab4eb74a70).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106922534

File: src/solvers/strings/string_constraint_generator_format.cpp [hunk 6]
```diff
@@ -365,6 +366,55 @@ add_axioms_for_format_specifier(
...
+      index_exprt{string.content(), from_integer(3, string.length().type())},
+      type};
+  }
+  if(id == ID_float)
+  {
+    // Deserialize an int and cast to float
+    const exprt as_int = format_arg_from_string(string, ID_int);
+    return typecast_exprt{as_int, floatbv_typet{}};
+  }
+  UNHANDLED_CASE;
```
Comment (peterschrammel) [inline src/solvers/strings/string_constraint_generator_format.cpp:434]: byte, short, double and long are not handled yet?

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/strings/string_constraint_generator_format.cpp (hunk 6 above)
Comment (romainbrenguier) [reply src/solvers/strings/string_constraint_generator_format.cpp:434]: They are all handled by being cast to a long within the model. So ID_int does not need to make a distinction between byte, short, int and long.

**Comment by romainbrenguier** [comment]
The java-models-library update is now pointing to master

File: src/solvers/strings/string_constraint_generator_format.cpp (hunk 6 above)
Comment (tautschnig) [reply src/solvers/strings/string_constraint_generator_format.cpp:434]: Drive-by comment: so what about `double`?

Comment (romainbrenguier) [reply src/solvers/strings/string_constraint_generator_format.cpp:434]: It's also cast to a long. ID_float handles both the cases where the formatting string except a float and a double (this was the same before my PR).

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7c34ea4258d32854669aa3dc94084ae37c93fd1e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106943324

**Pull request merged**